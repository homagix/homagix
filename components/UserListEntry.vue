<script setup lang="ts">
import { roles, type User } from "~/types"

const props = defineProps<{ user: User }>()

const users = await useUsers()

const formData = ref({
  role: props.user.role,
})

function saveUser() {
  users.update(props.user.id, formData.value)
}
</script>

<template>
  <td>{{ user.firstName }}</td>
  <td>{{ user.repository }}</td>
  <td>
    <select v-model="formData.role" @change="saveUser">
      <option v-for="role in roles" :selected="role === user.role">{{ role }}</option>
    </select>
  </td>
</template>

<style lang="scss" scoped>
select {
  font-size: 100%;
}
</style>
