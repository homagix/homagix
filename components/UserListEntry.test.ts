import { randomUUID } from "node:crypto"
import { describe, it, expect, vi } from "vitest"
import { setup } from "@nuxt/test-utils"
import { mountSuspended } from "@nuxt/test-utils/runtime"
import { mockNuxtImport } from "@nuxt/test-utils/runtime"
import type { User } from "~/types"
import UserListEntry from "./UserListEntry.vue"

const update = vi.fn()
mockNuxtImport("useUsers", () => {
  return () => ({ update })
})

const users = await useUsers()

const user: User = {
  id: randomUUID(),
  firstName: "arnold",
  repository: "https://github.com/arnold/t100",
  role: "author",
}

describe("UserListEntry", async () => {
  setup({ server: false })

  it("should render as expected", async () => {
    const wrapper = await mountSuspended(UserListEntry, { props: { user } })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should call the update function when the role is changed", async () => {
    users.update = vi.fn()
    const wrapper = await mountSuspended(UserListEntry, { props: { user } })
    await wrapper.find("select").setValue("admin")
    expect(update).toBeCalledWith(user.id, { role: "admin" })
  })
})
