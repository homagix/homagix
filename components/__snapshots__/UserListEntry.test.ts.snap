// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserListEntry > should render as expected 1`] = `
"<td data-v-769e4ae7="">arnold</td>
<td data-v-769e4ae7="">https://github.com/arnold/t100</td>
<td data-v-769e4ae7=""><select data-v-769e4ae7="">
    <option data-v-769e4ae7="">reader</option>
    <option data-v-769e4ae7="" selected="">author</option>
    <option data-v-769e4ae7="">admin</option>
  </select></td>"
`;
